docs/architecture/adr-006-light-client-refactor.md

```diff
@@ -0,0 +1,100 @@
+# ADR 006: Light Client Refactor
+
+## Changelog
+- 2020-03-15: Initial draft
+- 2020-03-23: New decomposition
+
+## Context
+
+The light client protocol provides a method for verifying application
+state without the execution of all preceding transactions. The
+protocol for the light client is described in
+[English](https://github.com/tendermint/spec/tree/bucky/light-reorg/spec/consensus/light)
+while the rust implementation is described in
+[ADR-002](adr-002-light-client-adr-index.md). This ADR outlines the
+next iteration of the light client implementation in rust in which
+multiple modules (Peer Exchange, Requester, Verifier, etc.) operate
+concurrently.
+
+The basis for this work comes from the learnings of the [Reactor
+Experiments](https://github.com/informalsystems/reactor-experiments) as
+well as the [Blockchain Reactor
+Refactor](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-043-blockchain-riri-org.md). The key take aways from that work are:
+
+1. Separate concerns into independently verifiable (by humans as well as
+   computers) components.
+2. Deterministic functions wherever possible
+3. Use events to index parameters of component function executions to
+   maintain alignment between specs and implementations.
+
+## Decision
+
+Model the Scheduler ,Verifier, Detector and IO as separate
+independently verifiable components. Communication between components is
+done by a demuxer component which ownes state and dispatches events
+between business logic executing components. Business logic is executed
+as a synchronous deterministic function. The composition of components
+can also be executed as a deterministic function.
+
+![Decomposition Diagram](assets/light-client-decomposition.png)
+
+### State
+The light client maintains state about commit/header/validators abbreviated as LightBlocks. It also has state pertaining to peers and evidence.
```

Comment:
LightBlocks will need to maintain ValidatorSet and NextValidatorSet.

Reply:
Out of curiosity: is there something wrong with using ValidatorSet (not NextValidatorSet)? _(other than validator set might have changed between ValidatorSet and NextValidatorSet)_